library/core/src/iter/adapters/fuse.rs

```diff
@@ -507,43 +422,6 @@ where
         P: FnMut(&Self::Item) -> bool,
         I: DoubleEndedIterator,
     {
-        unchecked!(self).rfind(predicate)
-    }
-
-    #[inline]
-    fn len(&self) -> usize
-    where
-        I: ExactSizeIterator,
-    {
-        unchecked!(self).len()
-    }
-
-    #[inline]
-    fn is_empty(&self) -> bool
-    where
-        I: ExactSizeIterator,
-    {
-        unchecked!(self).is_empty()
+        spec!(self.iter.rfind(predicate))
     }
 }
-
-#[unstable(issue = "none", feature = "inplace_iteration")]
-unsafe impl<S: Iterator, I: FusedIterator> SourceIter for Fuse<I>
-where
-    I: SourceIter<Source = S>,
-{
-    type Source = S;
-
-    #[inline]
-    unsafe fn as_inner(&mut self) -> &mut S {
-        match self.iter {
-            // SAFETY: unsafe function forwarding to unsafe function with the same requirements
-            Some(ref mut iter) => unsafe { SourceIter::as_inner(iter) },
-            // SAFETY: the specialized iterator never sets `None`
-            None => unsafe { intrinsics::unreachable() },
```

Comment:
The trait impl is being removed due to the unreachable here, right? I think it might be possible to replace this with a panic instead.

If someone pulled the HRTB shenaningans then it would only lead to a panic, not to unsafety. If that's not good enough then maybe `SourceIter` could be changed to return an `Option<&mut>`.

Reply:
I personally don't think we should use a panic for code that is otherwise allowable, however weird it may be.

Changing `SourceIter` to return an `Option` sounds promising, assuming there actually is a reasonable fallback at the caller -- I suppose going back to normal not-in-place collection. Then you wouldn't even need the seemingly-unrelated `I: FusedIterator` constraint here.

Reply:
>  I suppose going back to normal not-in-place collection. 

That would probably be bad for compile times since it would have to add two different code paths in that case.

If we can assume that `None` implies an exhausted iterator then it could work and simply return an empty Vec. But I'll have to look through all the implementations to see if that's a valid assumption.

Reply:
> Then you wouldn't even need the seemingly-unrelated I: FusedIterator constraint here.

In place iteration works roughly like this

1. retrieve source iter to get a write pointer
2. perform loop that gets elements through tie iterators and writes to the pointer
3. retrieve source iter again to drop the remainder and then steal the allocation

Which means that once we have started iterating the source must stay retrievable. Which means the iterator cannot be fused through the option because that would drop the source. So it can only work for `FusedIterator` which doesn't switch the option variant.

Under those circumstances it should be safe since user code doesn't have access to the iterator during that time and can't do the variance coercion.

Reply:
OK, yeah, you need it to not drop. I agree the user doesn't have access during your steps, but they could have played the variance game before it got there at all, so it might be `None` to start with. Then an empty collection is fine.